recourse_methods/base_type.py

```diff
@@ -22,14 +22,19 @@ def get_all_recourse_directions(
         """
 
     @abc.abstractmethod
-    def get_all_recourse_instructions(self, poi: pd.Series) -> Sequence[Any]:
+    def get_all_recourse_instructions(
+        self, poi: pd.Series
+    ) -> Sequence[Optional[Any]]:
         """Generates different recourse instructions for the poi for each of
         the k_directions.
 
         Whereas recourse directions are vectors in embedded space,
         instructions are human-readable guides for how to follow those
         directions in the original data space.
 
+        It always returns a sequence of length k. If no recourse is possible,
```

Comment:
Do you need the failure mode to return a sequence of length `k`? Why not just a sequence of a single `None` element?

Reply:
This function returns multiple instructions (one for each path), so it can partially fail by, for example, finding 2 recourse instructions instead of 3. It seemed too sneaky to return a list of 2 instructions when the user expects 3. Returning Nones is more likely to throw obvious and immediate errors in user code than returning a shorter-than-expected list.